src/proxy/metrics.rs

```diff
@@ -449,35 +459,54 @@ impl ConnectionResult {
 
             "connection opened"
         );
+        // Grab the metrics with our labels now, so we don't need to fetch them each time.
+        // The inner metric is an Arc so clone is fine/cheap.
+        // With the raw Counter, we increment is a simple atomic add operation (~1ns).
+        // Fetching the metric itself is ~300ns; fast, but we call it on each read/write so it would
+        // add up.
+        let sent_metric = metrics.sent_bytes.get_or_create(&tl).clone();
+        let recv_metric = metrics.received_bytes.get_or_create(&tl).clone();
+        let sent = atomic::AtomicU64::new(0);
+        let recv = atomic::AtomicU64::new(0);
         Self {
             src,
             dst,
             hbone_target,
             start,
             tl,
             metrics,
+
+            sent,
+            sent_metric,
+            recv,
+            recv_metric,
         }
     }
 
     pub fn increment_send(&self, res: u64) {
-        let tl = &self.tl;
-        self.metrics.sent_bytes.get_or_create(tl).inc_by(res);
+        self.sent.inc_by(res);
+        self.sent_metric.inc_by(res);
     }
+
     pub fn increment_recv(&self, res: u64) {
-        let tl = &self.tl;
-        self.metrics.received_bytes.get_or_create(tl).inc_by(res);
+        self.recv.inc_by(res);
+        self.recv_metric.inc_by(res);
     }
+
     // Record our final result.
     // Ideally, we would save and report from the increment_ functions instead of requiring a report here.
-    pub fn record<E: std::error::Error>(&self, res: Result<(u64, u64), E>) {
+    pub fn record<E: std::error::Error>(&self, res: Result<(), E>) {
         let tl = &self.tl;
 
         // Unconditionally record the connection was closed
         self.metrics.connection_close.get_or_create(tl).inc();
 
         // Unconditionally write out an access log
         let mtls = tl.connection_security_policy == SecurityPolicy::mutual_tls;
-        let bytes = res.as_ref().ok();
+        let bytes = (
+            self.recv.load(Ordering::SeqCst),
```

Comment:
Presumably this would be much slower than the writes, but is out of the hotpath so should be OK.

Reply:
On x86 Relaxed is just a `mov` (read: atomic is ~free). SeqCst is `xchg`. Still very cheap, and we do this once per connection -- this is actually probably the cheapest line in this function